# Task-1: Data Quality Issues _ Transactions

The client has provided us with four datasets.
They are:
1) Customer Demographic
2) Transaction details
3) Customer address
4) New customer list

In [1]:
#Import the required librariesabs
import pandas as pd

## Reading Excel file given by the client

In [2]:
df=pd.read_csv("G:\My Drive\KPMG\Task 1\Transactions.csv")

Let us copy the same data frame (as below) so that changes to be made in future may not affect the Excel file

In [3]:
df=df.copy()

## Viewing the data frame

In [4]:
df.head(5)

transaction_id  product_id  customer_id transaction_date online_order  \
0            8708           0         5034        7/10/2017        False   
1           16701           0         5034       27-01-2017        False   
2           17469           0         5034         3/1/2017        False   
3            2463           2         3500       25-04-2017        False   
4            6309          69         3500         9/1/2017         True   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex         Road        medium       medium   
1     Approved  Norco Bicycles     Standard        medium       medium   
2     Approved      OHM Cycles         Road        medium       medium   
3     Approved           Solex     Standard        medium       medium   
4     Approved  Giant Bicycles         Road        medium       medium   

   list_price standard_cost product_first_sold_date  
0      416.98   312.7350159                2/8/2015  
1      360.40   270.2999878              25-08-1997  
2      742.54   667.4000244               2/10/1992  
3       71.49       $53.62                2/12/2012  
4      792.90      $594.68                2/10/1992

## Exploring shape of the sheet 

In [5]:
df.shape

(20000, 13)

## Exploring columns of the sheets 

In [6]:
df.columns

Index(['transaction_id', 'product_id', 'customer_id', 'transaction_date',
       'online_order', 'order_status', 'brand', 'product_line',
       'product_class', 'product_size', 'list_price', 'standard_cost',
       'product_first_sold_date'],
      dtype='object')

## Checking Data Types of each attribute in each sheet

In [7]:
df.dtypes

transaction_id               int64
product_id                   int64
customer_id                  int64
transaction_date            object
online_order                object
order_status                object
brand                       object
product_line                object
product_class               object
product_size                object
list_price                 float64
standard_cost               object
product_first_sold_date     object
dtype: object

## Describe the data columnwise use "df.describe()"

## Checking for Null/missing values

In [8]:
df.isnull().sum()

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               360
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

## Exploring columns 

In [9]:
df.iloc[:,7].value_counts()

Standard    14176
Road         3970
Touring      1234
Mountain      423
Name: product_line, dtype: int64

Here, in [], : before comma indicates the row index, here : means all rows. Similarly values after comma indicates the column index, here 7 means 8th column Because, the indices in python start with zero.

Below is the other way to explore the same column, i.e. Transactions.iloc[:,7] = product_line

In [10]:
df['product_line'].value_counts()

Standard    14176
Road         3970
Touring      1234
Mountain      423
Name: product_line, dtype: int64

## Why to see value_counts seperately for each column?, Make a loop so that together we get the desired response for all columns.. 

In [11]:
df.columns

Index(['transaction_id', 'product_id', 'customer_id', 'transaction_date',
       'online_order', 'order_status', 'brand', 'product_line',
       'product_class', 'product_size', 'list_price', 'standard_cost',
       'product_first_sold_date'],
      dtype='object')

In [12]:
df_columns=['transaction_id', 'product_id', 'customer_id', 'transaction_date',
       'online_order', 'order_status', 'brand', 'product_line',
       'product_class', 'product_size', 'list_price', 'standard_cost',
       'product_first_sold_date']

In [13]:
for i in df_columns:
    print(df[i].value_counts())

2047     1
10912    1
12947    1
2708     1
661      1
6806     1
4759     1
19100    1
17053    1
8865     1
8849     1
15010    1
12963    1
2724     1
677      1
6822     1
4775     1
19116    1
14994    1
10896    1
12787    1
10880    1
12915    1
2676     1
629      1
6774     1
4727     1
19068    1
17021    1
8833     1
        ..
15693    1
11599    1
19747    1
17746    1
19795    1
5464     1
7513     1
1370     1
3419     1
13660    1
17730    1
11583    1
9534     1
15677    1
5416     1
7465     1
1322     1
3371     1
13612    1
15661    1
9518     1
11567    1
17714    1
19763    1
5432     1
7481     1
1338     1
3387     1
13628    1
2049     1
Name: transaction_id, Length: 20000, dtype: int64
0      1378
3       354
1       311
35      268
38      267
4       241
2       240
90      225
12      224
80      223
5       222
43      216
69      215
77      215
27      215
53      214
45      213
92      211
36      211
50      209
64      208
40      208
79      207
17 

In [14]:
df["product_first_sold_date"]= pd.to_datetime(df["product_first_sold_date"])

In [15]:
df["product_first_sold_date"]

0       2015-02-08
1       1997-08-25
2       1992-02-10
3       2012-02-12
4       1992-02-10
5       2000-05-22
6       1992-11-10
7       1993-07-15
8       2006-02-02
9       2010-07-06
10      2003-02-16
11      1998-12-17
12      2012-02-12
13      1993-06-23
14      2006-02-02
15      2006-05-22
16      2003-02-16
17      1999-07-26
18      2013-09-16
19      2004-08-17
20      1994-10-08
21      2004-01-16
22      2016-11-14
23      1996-05-04
24      2004-07-25
25      2014-10-10
26      1994-12-07
27      2003-07-02
28      2010-05-05
29      1997-08-25
           ...    
19970   2008-03-19
19971   1993-02-10
19972   2003-05-01
19973   2008-03-19
19974   1991-06-05
19975   2009-08-03
19976   2003-10-09
19977   1993-06-23
19978   2012-10-04
19979   2012-10-04
19980   2014-03-03
19981   2000-03-11
19982   2010-05-05
19983   2002-08-31
19984   2010-05-05
19985   1998-12-16
19986   2011-08-29
19987   2004-07-08
19988   2002-03-22
19989   2005-10-05
19990   2008-03-19
19991   2003

## Removing dollar sign from the column regarding cost 

In [16]:
df['standard_cost']=df['standard_cost'].str.replace('$','')
df['standard_cost'].head()

0    312.7350159
1    270.2999878
2    667.4000244
3         53.62 
4        594.68 
Name: standard_cost, dtype: object

In [17]:
df['standard_cost'].dtype

dtype('O')

Though dollar sign is successfully removed, data type is not float. Need to convert it into float. 

In [18]:
df['standard_cost']=df['standard_cost'].astype(float)
df['standard_cost']

ValueError: could not convert string to float: '1,167.18 '

Data could not be converted to float.. because of "ValueError: could not convert string to float: '1,167.18'"
So here, ',' need to be removed.

In [19]:
df['standard_cost']=df['standard_cost'].str.replace(',','')
df['standard_cost'].head()

0    312.7350159
1    270.2999878
2    667.4000244
3         53.62 
4        594.68 
Name: standard_cost, dtype: object

Good. now let us try toconvers it to float type.

In [20]:
#df['standard_cost'].astype(float)

## Determining percetage of  missing values in each column.

In [21]:
missing_count=df.isnull().sum()
missing_count

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               360
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

In [22]:
total=len(df)
total

20000

In [23]:
percentageMissing=missing_count*100/total
percentageMissing

transaction_id             0.000
product_id                 0.000
customer_id                0.000
transaction_date           0.000
online_order               1.800
order_status               0.000
brand                      0.985
product_line               0.985
product_class              0.985
product_size               0.985
list_price                 0.000
standard_cost              0.985
product_first_sold_date    0.985
dtype: float64

In [24]:
missing_data=pd. DataFrame({'Heading': df.columns, 'total_missing': missing_count,'Percentage_missing': percentageMissing})
missing_data.sort_values('total_missing', ascending=False)
missing_data=missing_data.loc[missing_data['total_missing'] > 0]
missing_data

Heading  total_missing  \
online_order                        online_order            360   
brand                                      brand            197   
product_line                        product_line            197   
product_class                      product_class            197   
product_size                        product_size            197   
standard_cost                      standard_cost            197   
product_first_sold_date  product_first_sold_date            197   

                         Percentage_missing  
online_order                          1.800  
brand                                 0.985  
product_line                          0.985  
product_class                         0.985  
product_size                          0.985  
standard_cost                         0.985  
product_first_sold_date               0.985

## Deteriming if duplicate value is there in the data frame

In [25]:
duplicate=df[df.duplicated()]
duplicate

Empty DataFrame
Columns: [transaction_id, product_id, customer_id, transaction_date, online_order, order_status, brand, product_line, product_class, product_size, list_price, standard_cost, product_first_sold_date]
Index: []

In [26]:
len(duplicate)

0

## Viewing categorical data

In [27]:
df['online_order'].value_counts()

True     9829
False    9811
Name: online_order, dtype: int64

In [28]:
df['order_status'].value_counts()

Approved     19821
Cancelled      179
Name: order_status, dtype: int64

In [29]:
df['brand'].value_counts()

Solex             4253
Giant Bicycles    3312
WeareA2B          3295
OHM Cycles        3043
Trek Bicycles     2990
Norco Bicycles    2910
Name: brand, dtype: int64

In [30]:
df['product_class'].value_counts()

medium    13826
high       3013
low        2964
Name: product_class, dtype: int64

In [31]:
df['product_line'].value_counts()

Standard    14176
Road         3970
Touring      1234
Mountain      423
Name: product_line, dtype: int64

In [32]:
df['product_class'].value_counts()

medium    13826
high       3013
low        2964
Name: product_class, dtype: int64

In [33]:
df['product_size'].value_counts()

medium    12990
large      3976
small      2837
Name: product_size, dtype: int64

## Converting object data type into date format 

In [34]:
df["product_first_sold_date"]= pd.to_datetime(df["product_first_sold_date"])

In [35]:
df["product_first_sold_date"].head()

0   2015-02-08
1   1997-08-25
2   1992-02-10
3   2012-02-12
4   1992-02-10
Name: product_first_sold_date, dtype: datetime64[ns]

In [36]:
df["transaction_date"]= pd.to_datetime(df["transaction_date"])

In [37]:
df["product_first_sold_date"].head()

0   2015-02-08
1   1997-08-25
2   1992-02-10
3   2012-02-12
4   1992-02-10
Name: product_first_sold_date, dtype: datetime64[ns]

## Save the data frame in '.csv' fromat

In [38]:
df.to_csv('transaction_ready.csv', index=False)